satellite/satellitedb/overlaycache.go

```diff
@@ -22,6 +24,243 @@ type overlaycache struct {
 	db *dbx.DB
 }
 
+type filterReputableRequest struct {
+	minReputation         *pb.NodeStats
+	freeBandwidth         int64
+	freeDisk              int64
+	excluded              []pb.NodeID
+	reputableNodeAmount   int64
+	newNodeAuditThreshold int64
+}
+
+type filterNewRequest struct {
+	freeBandwidth         int64
+	freeDisk              int64
+	excluded              []pb.NodeID
+	newNodeAmount         int64
+	newNodeAuditThreshold int64
+}
+
+// FilterNodes looks up nodes based on reputation requirements
+func (cache *overlaycache) FilterNodes(ctx context.Context, req *overlay.FilterNodesRequest) ([]*pb.Node, error) {
+	reputableNodeAmount := req.MinNodes
+	if reputableNodeAmount <= 0 {
+		reputableNodeAmount = req.Opts.GetAmount()
+	}
+
+	filterReputableReq := &filterReputableRequest{
+		minReputation:         req.MinReputation,
+		freeBandwidth:         req.Opts.GetRestrictions().FreeBandwidth,
+		freeDisk:              req.Opts.GetRestrictions().FreeDisk,
+		excluded:              req.Opts.ExcludedNodes,
+		reputableNodeAmount:   reputableNodeAmount,
+		newNodeAuditThreshold: req.NewNodeAuditThreshold,
+	}
+
+	reputableNodes, err := cache.filterReputableNodes(ctx, filterReputableReq)
+	if err != nil {
+		return nil, err
+	}
+
+	newNodeAmount := reputableNodeAmount * int64(float64(reputableNodeAmount)*req.NewNodePercentage)
+
+	filterNewReq := &filterNewRequest{
+		freeBandwidth:         req.Opts.GetRestrictions().FreeBandwidth,
+		freeDisk:              req.Opts.GetRestrictions().FreeDisk,
+		excluded:              req.Opts.ExcludedNodes,
+		newNodeAmount:         newNodeAmount,
+		newNodeAuditThreshold: req.NewNodeAuditThreshold,
+	}
+
+	newNodes, err := cache.filterNewNodes(ctx, filterNewReq)
+	if err != nil {
+		return nil, err
+	}
+
+	var allNodes []*pb.Node
+	allNodes = append(allNodes, reputableNodes...)
+	allNodes = append(allNodes, newNodes...)
+
+	return allNodes, nil
+}
+
+func (cache *overlaycache) filterReputableNodes(ctx context.Context, req *filterReputableRequest) (reputableNodes []*pb.Node, err error) {
+	rows, err := cache.findReputableNodesQuery(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err = utils.CombineErrors(err, rows.Close())
+	}()
+
+	for rows.Next() {
```

Comment:
this section as far as I can tell is identical for both `filterReputableNodes` and `filterNewNodes`, so maybe it's worth pulling out into a helper function like `sqlRowsToDbxNode` or something.